### PR TITLE
Newsletter signup

### PR DESCRIPTION
**Pull request by davidpmccormick**
References #2557

Adding a newsletter signup page at `/info/newsletter`.

__TODO__
- [ ] Copy
- [ ] Style
- [x] Front-end validation (?HTML5 validations maybe sufficient)
- [x] Handle any errors returned from dotmailer (sort of – for now we're just saying that there was a problem)

__1. `/info/newsletter`__
screen shot 2018-05-10 at 12 02 23

__2. `/info/newsletter` after attempted submit with client side validation errors__
screen shot 2018-05-10 at 12 04 23

__3. `/info/newsletter?result=success` after successful form submission__
screen shot 2018-05-10 at 11 53 59

__4. `/info/newsletter?result=error` after receiving an error from dotmailer__
screen shot 2018-05-10 at 12 05 44

### CODE REVIEW

File: client/js/components/newsletter-signup.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  el.setAttribute('novalidate', '');
+
+  function handleSubmit(event) {
+    event.preventDefault();
+
+    isSubmitAttempted = true;
+
+    if (!emailInputEl.validity.valid) {
+      el.classList.add('is-error');
+      el.append(emailErrorMessage);
```
Comment (jamesgorrie) [inline client/js/components/newsletter-signup.js]: We could just use browser default error handling for this?

File: common/views/components/NewsletterSignup/NewsletterSignup.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        {!isConfirmed && !isSuccess &&
+          <form
+            noValidate={this.state.noValidate}
+            onSubmit={this.handleSubmit}
+            name='newsletter-signup'
+            id='newsletter-signup'
+            action='https://r1-t.trackedlink.net/signup.ashx'
+            method='post'>
+            <input type='hidden' name='userid' value='225683' />
+            <input type='hidden' name='ReturnURL' value='http://localhost:3000/info/newsletter' />
```
Comment (jamesgorrie) [inline common/views/components/NewsletterSignup/NewsletterSignup.js]: ?

File: common/views/components/NewsletterSignup/NewsletterSignup.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            <input type='hidden' name='SIG22a9ece3ebe9b2e10e328f234fd10b3f5686b9f4d45f628f08852417032dc990' value='' />
```
Comment (jamesgorrie) [inline common/views/components/NewsletterSignup/NewsletterSignup.js]: ?

File: common/views/components/NewsletterSignup/NewsletterSignup.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+        {!isConfirmed && !isSuccess &&
+          <form
+            noValidate={this.state.noValidate}
+            onSubmit={this.handleSubmit}
+            name='newsletter-signup'
+            id='newsletter-signup'
+            action='https://r1-t.trackedlink.net/signup.ashx'
+            method='post'>
+            <input type='hidden' name='userid' value='225683' />
```
Comment (jamesgorrie) [inline common/views/components/NewsletterSignup/NewsletterSignup.js]: ?

File: common/views/components/NewsletterSignup/NewsletterSignup.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+              <p className={`${spacing({s: 2}, {padding: ['top', 'right', 'bottom', 'left'], margin: ['bottom']})} border-width-1 border-color-red font-red`}>Please enter a valid email address.</p>
+            }
+          </form>
+        }
+      </div>
+    );
+  }
+}
+
+export default NewsletterSignup;
```
Comment (jamesgorrie) [inline common/views/components/NewsletterSignup/NewsletterSignup.js]: I'm a little confused as to why we have both the React client handling as well as the `client/app.js` handling?
It makes it quite confusing as to what is what, and it seems like a lot of testing code has made it in?

File: server/routes/index.js [hunk 6]
```diff
@@ -56,6 +57,7 @@ r.get('/series/(W):id', renderSeries);
 r.get('/webcomic-series/:id', renderWebcomicSeries);
 r.get('/info/opening-times', renderOpeningTimes);
 r.get('/pages/:id', renderPage);
+r.get('/info/newsletter', renderNewsletterPage);
```
Comment (jamesgorrie) [inline server/routes/index.js]: I think we might want to chat about `/info` pages, as they don't exist any more.

File: client/js/components/newsletter-signup.js (hunk 1 above)
Comment (davidpmccormick) [reply client/js/components/newsletter-signup.js]: We use browser default error handling, but progressively enhance it.

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 2 above)
Comment (davidpmccormick) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: Yep – needs changing! Thanks.

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 3 above)
Comment (davidpmccormick) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: dotmailer needs this (and provided us the name)

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 4 above)
Comment (davidpmccormick) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: this is Wellcome Collection's dotmailer account

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 5 above)
Comment (davidpmccormick) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: We'll need the component to work once we start using it in React, so felt like it made sense to write equivalent validation code for the React and non-React versions at the same time.

**Comment by davidpmccormick** [comment]
> Just wondering what seems to be a lot of testing code is?

Do you mean the form validation? There's a client-side check for a valid(ish) email address and at least one checkbox being checked.

Or was it something else?

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 3 above)
Comment (jamesgorrie) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: Fancy leaving a note as it looks like testing code?

File: common/views/components/NewsletterSignup/NewsletterSignup.js (hunk 4 above)
Comment (jamesgorrie) [reply common/views/components/NewsletterSignup/NewsletterSignup.js]: Perhaps a note?

**Comment by jamesgorrie** [comment]
re: React - it might be good to do them as separate reviews, as things like form / state / serverside strategy questions are going to come up (two covered in this PR) and it would be good to approach them with the thought of the decisions being things we will try to implement in the future.

**Pull request merged**